// gmdf/lib/web/services/index_service.js
var mysql_utils = require('../../common/utils/mysql_utils')();

// 根据登陆账号获取用户信息
exports.queryUserByAccountOrPhone = function(login_account) {

    var sql = 'select a.* from common_user_info a where a.login_account = ? or a.user_phone = ? ';

    return mysql_utils.find(sql, [login_account, login_account]);
}

// 根据登陆账号获取用户信息
exports.queryUserByAccountOrPhone = function(login_account) {

    var sql = 'select a.*, b.sys_theme_layout, b.sys_main_url from common_user_info a left join common_system_info b on a.user_sys = b.id where a.login_account = ? or a.user_phone = ? ';

    return mysql_utils.find(sql, [login_account, login_account]);
}

// 查询系统导航
exports.querySysNav = function(sys_id, user_id) {

    var sql = 'select a.* from common_menu_info a where a.menu_sysid = ? and a.menu_level = 1 and a.menu_status = 1 and a.id in (select distinct m.menu_id from common_role_menu_opt m left join common_user_roles n on m.role_id = n.role_id where n.user_id = ?)  order by a.menu_order';
    return mysql_utils.find(sql, [sys_id, user_id]);
}

// 查询系统菜单
exports.querySysMenu = function(sys_id, menu_pid, user_id) {
    var sql = 'select a.id as id, a.menu_pid as pid, a.menu_code as code, a.menu_nav as nav, a.menu_name as title, a.menu_cls as icon, a.menu_url as href, a.menu_target as target from common_menu_info a where a.menu_sysid = ? and a.menu_level > 1 and a.menu_status = 1 and a.menu_hidden = 0 and (a.menu_pid = ? or a.menu_pid in (select t.id from common_menu_info t where t.menu_pid = ?)) and a.id in (select distinct m.menu_id from common_role_menu_opt m left join common_user_roles n on m.role_id = n.role_id where n.user_id = ?) order by a.menu_pid, a.menu_order';
    return mysql_utils.find(sql, [sys_id, menu_pid, menu_pid, user_id]);
}

// 查询用户信息
exports.queryUserById = function(user_id) {
    var sql = 'select a.login_account,a.login_password, a.user_name, a.user_no, a.user_phone, a.user_email, d.org_name, GROUP_CONCAT(c.role_name) as user_roles, get_dict_text(\'common_user_type\', a.user_type) as user_type_name FROM common_user_info a left join common_user_roles b on a.user_id = b.user_id left join common_role_info c on b.role_id = c.id left join common_org_info d on a.user_org = d.org_id where a.user_id = ?';
    return mysql_utils.find(sql, [user_id]);
}
// 修改用户密码
exports.updateUserPwd = function(user_id, user_pwd) {
    return mysql_utils.update('common_user_info', {user_id : user_id}, {login_password : user_pwd});
}

// 查询用户角色拥有的权限
exports.getUserRoleMenus = function(user_id) {
    var sql = 'select distinct d.id, d.menu_url, d.menu_level from common_user_roles a left join common_role_info b on a.role_id = b.id left join common_role_menu_opt c on b.id = c.role_id left join common_menu_info d on c.menu_id = d.id where b.role_status = 1 and c.type = 1 and d.menu_status = 1 and a.user_id = ? ';
    return mysql_utils.find(sql, [user_id]);
}

// 查询用户职责
exports.queryUserDuties = function(user_id) {

    var sql = 'select a.* from common_user_duties a where a.user_id = ? order by duty_type, duty_id';

    return mysql_utils.find(sql, [user_id]);
}

// 记录用户登陆日志
exports.saveUserLoginLog = function(params) {
    return mysql_utils.save('common_user_log_login', params);
}


// 记录用户登陆错误日志
exports.saveUserLoginErrorLog = function(params) {
    return mysql_utils.save('common_user_log_login_error', params);
}

/**
 * 查找所有系统
 */
exports.querySysList = function(sys_status){
    if(sys_status != 0) {
        sys_status = 1;
    }
    var sql="select a.id, a.sys_code, a.sys_name from common_system_info a where a.sys_status = ?";
    return mysql_utils.find(sql, [sys_status]);
};

/**
 * 查询参数类别
 * @param sysid
 * @param catalogcode
 */
exports.querySysParamCatalog = function(sysid, catalogcode) {
    // 查询类别信息
    var sql = "select a.* from common_param_catalog_info a where a.catalog_status = 1 and a.catalog_sysid = ? and a.catalog_code = ?";
    return mysql_utils.find(sql, [sysid, catalogcode]);
}

/**
 * 查询系统参数
 * @param catalogid
 */
exports.querySysParam = function(catalogid) {
    // 查询类别信息
    var sql = "select a.* from common_param_info a where a.param_status = 1 and a.param_catalogid = ? order by a.param_id";
    return mysql_utils.find(sql, [catalogid]);
}

/**
 * 修改系统参数
 * @param param_id
 * @param param_val
 * @returns {null}
 */
exports.updateSysParamValue = function(param_id, param_val) {
    return mysql_utils.update('common_param_info', {param_id:param_id}, {param_val:param_val});
}
