// gmdf/lib/web/ecl/dict.js
var desp = {
    app_id: 'gmdf-dict',// 模块编码
    app_author: 'yanghang',// 模块开发人员，gitlib账号名
    app_issueid:'', //开发任务ID ，如 #128
    app_biz_map: {// 功能列表
        '10': {// 功能编号
            app_biz_desp: '字典列表', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查询字典列表数据成功', desp: '查询字典列表数据成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查询字典列表数据异常', desp: '查询字典列表数据异常'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '11': {// 功能编号
            app_biz_desp: '保存字典', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '保存字典信息成功', desp: '保存字典信息成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '保存字典信息异常', desp: '保存字典信息异常'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '12': {// 功能编号
            app_biz_desp: '修改字典信息', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '修改字典信息成功', desp: '修改字典信息成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '必须选中一条数据', desp: 'id不能为空'},
                5101: {sdesp: '修改字典信息失败', desp: '修改字典信息失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '13': {// 功能编号
            app_biz_desp: '获取字典的属性列表', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查找成功', desp: '查找成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查找字典属性列表失败', desp: '查找字典属性列表失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '14': {// 功能编号
            app_biz_desp: '所有属性列表', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查找成功', desp: '查找成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查找所有字典属性列表失败', desp: '查找所有字典属性列表失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '15': {// 功能编号
            app_biz_desp: '保存字典属性信息', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '保存字典属性成功', desp: '保存字典属性成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '保存字典属性失败', desp: '保存字典属性失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '16': {// 功能编号
            app_biz_desp: '修改字典属性信息', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '修改字典属性成功', desp: '修改字典属性成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '必须选中一条数据', desp: 'id不能为空'},
                5101: {sdesp: '修改字典属性失败', desp: '修改字典属性失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '17': {// 功能编号
            app_biz_desp: '获取上级属性列表', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '获取上级属性列表成功', desp: '获取上级属性列表成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '获取上级属性列表失败', desp: '获取上级属性列表失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '18': {// 功能编号
            app_biz_desp: '获取属性树形目录', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '获取属性树形目录成功', desp: '获取属性树形目录成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '获取属性树形目录失败', desp: '获取属性树形目录失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        }
    }
};

module.exports = desp;
