// gmdf/lib/web/ecl/index.js
var desp = {
    app_id: 'gmdf-core',// 模块编码
    app_author: 'shihukui',// 模块开发人员，gitlib账号名
    app_issueid:'', //开发任务ID ，如 #128
    app_biz_map: {// 功能列表
        '10': {// 功能编号
            app_biz_desp: '登陆', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '登陆成功', desp: '登陆成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '登陆账号不能为空', desp: '登陆账号不能为空'},
                5101: {sdesp: '登陆密码不能为空', desp: '登陆密码不能为空'},
                5102: {sdesp: '验证码不能为空', desp: '验证码不能为空'},
                5103: {sdesp: '登录时出现错误', desp: '登录时出现错误'},
                5104: {sdesp: '用户名或密码错误', desp: '密码错误'},
                5105: {sdesp: '用户名或密码错误', desp: '账号重复'},
                5106: {sdesp: '用户名或密码错误', desp: '账号不存在'},
                5107: {sdesp: '验证码不正确', desp: '验证码不正确'},
                5108: {sdesp: '账号已被冻结', desp: '账号已被冻结'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
                5200: {sdesp: '保存参数失败', desp: '系统参数描述长度超出限制'}
            }
        },
        '11': {// 功能编号
            app_biz_desp: '修改密码', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                2000: {sdesp: '修改密码成功', desp: '修改密码成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '原始密码不能为空', desp: '原始密码不能为空'},
                5101: {sdesp: '新密码不能为空', desp: '新密码不能为空'},
                5102: {sdesp: '确认密码不能为空', desp: '确认密码不能为空'},
                5103: {sdesp: '两次输入密码不一致，请重新输入', desp: '新密码和确认密码不一致'},
                5104: {sdesp: '密码为8~18位大小写字母数字和特殊字符组成', desp: '新密码格式不正确'},
                5105: {sdesp: '原始密码错误', desp: '原始密码错误'},
                5106: {sdesp: '修改密码时出现错误', desp: '修改密码时出现错误'},
                5107: {sdesp: '修改密码时出现错误', desp: '修改密码时出现错误'}
            }
        }
        ,
        '12': {// 功能编号
            app_biz_desp: '系统列表', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                2000: {sdesp: '查询系统列表成功', desp: '查询系统列表成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查询系统列表失败', desp: '查询系统列表失败'}
            }
        }
        ,
        '13': {// 功能编号
            app_biz_desp: '更新系统参数值', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                2000: {sdesp: '保存系统参数成功', desp: '保存系统参数成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '保存系统参数失败', desp: '保存系统参数失败'},
                5101: {sdesp: '保存系统参数失败', desp: '保存系统参数失败,预期数和实际数不一致'}
            }
        }
    }
};

module.exports = desp;
