// gmdf/lib/web/ecl/menu.js
var desp = {
    app_id: 'gmdf-menu',// 模块编码
    app_author: 'yanghang',// 模块开发人员，gitlib账号名
    app_issueid:'', //开发任务ID ，如 #128
    app_biz_map: {// 功能列表
        '10': {// 功能编号
            app_biz_desp: '菜单列表', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查询菜单列表数据成功', desp: '查询菜单列表数据成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查询菜单列表数据异常', desp: '查询菜单列表数据异常'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '11': {// 功能编号
            app_biz_desp: '保存菜单', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '保存菜单成功', desp: '保存菜单成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '保存菜单数据异常', desp: '保存菜单数据异常'},
                5101: {sdesp: '有未填入的选项', desp: '有未填入的选项'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '12': {// 功能编号
            app_biz_desp: '修改菜单', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '修改菜单成功', desp: '修改菜单成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '必须选中一条数据', desp: 'id不能为空'},
                5101: {sdesp: '上级菜单与本菜单冲突', desp: '除了root菜单，上级菜单选了本菜单'},
                5102: {sdesp: '有未填入的选项', desp: '有未填入的选项'},
                5103: {sdesp: '修改菜单失败', desp: '修改菜单失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '13': {// 功能编号
            app_biz_desp: '查找菜单列表', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查找成功', desp: '查找成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查找菜单列表失败', desp: '查找菜单列表失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '14': {// 功能编号
            app_biz_desp: '获取系统列表', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查找成功', desp: '查找系统列表成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查找系统列表失败', desp: '查找系统列表失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '15': {// 功能编号
            app_biz_desp: '获取菜单的权限列表', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查找成功', desp: '查找系统列表成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查询菜单列表数据异常', desp: '查询菜单列表数据异常'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '16': {// 功能编号
            app_biz_desp: '保存菜单权限', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '保存菜单权限成功', desp: '保存菜单权限成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '保存菜单权限数据异常', desp: '保存菜单权限数据异常'},
                5101: {sdesp: '有未填入的选项', desp: '有未填入的选项'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '17': {// 功能编号
            app_biz_desp: '修改菜单权限', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '修改菜单权限成功', desp: '修改菜单权限成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '修改菜单权限失败', desp: '修改菜单权限失败'},
                5101: {sdesp: '有未填入的选项', desp: '有未填入的选项'},
                5102: {sdesp: '未选中数据', desp: 'id不能为空'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '18': {// 功能编号
            app_biz_desp: '批量保存菜单权限', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '生成菜单权限成功', desp: '生成菜单权限成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '生成菜单权限失败', desp: '生成菜单权限失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '19': {// 功能编号
            app_biz_desp: '生成操作中获取路由列表', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '读取路由列表成功', desp: '读取路由列表成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '没有找到相应的菜单路由', desp: '没有找到相应的菜单路由'},
                5101: {sdesp: '读取路由列表失败', desp: '读取路由列表失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '20': {// 功能编号
            app_biz_desp: '生成菜单树形目录', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '生成菜单树成功', desp: '生成菜单树成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '生成菜单树失败', desp: '生成菜单树失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '21': {// 功能编号
            app_biz_desp: '点击菜单查找子节点', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查找成功', desp: '查询节点及子节点信息成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查询失败', desp: '查询节点及子节点信息失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '22': {// 功能编号
            app_biz_desp: '切换系统', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查找成功', desp: '查询选中系统菜单信息成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查询失败', desp: '查询选中系统菜单信息失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '23': {// 功能编号
            app_biz_desp: '切换系统菜单目录', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查找成功', desp: '查询选中系统菜单目录成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查询失败', desp: '查询选中系统菜单目录失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        }
    }
};

module.exports = desp;
