// gmdf/lib/web/ecl/organization.js
var desp = {
    app_id: 'gmdf-organization',// 模块编码
    app_author: 'yangjingya',// 模块开发人员，gitlib账号名
    app_issueid:'', //开发任务ID ，如 #128
    app_biz_map: {// 功能列表
        '10': {// 功能编号
            app_biz_desp: '组织列表', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查询组织列表数据成功', desp: '查询组织列表数据成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查询组织列表数据异常', desp: '查询组织列表数据异常'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '11': {// 功能编号
            app_biz_desp: '保存组织', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '保存组织成功', desp: '保存组织成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '保存组织数据异常', desp: '保存组织数据异常'},
                5101: {sdesp: '输入数据有空缺', desp: '输入数据有空缺'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '12': {// 功能编号
            app_biz_desp: '修改组织', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '修改组织成功', desp: '修改组织成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '必须选中一条数据', desp: 'id不能为空'},
                5101: {sdesp: '上级组织与本组织冲突', desp: '除了root菜单，上级菜单选了本菜单'},
                5102: {sdesp: '输入数据有空缺', desp: '输入数据有空缺'},
                5103: {sdesp: '修改组织失败', desp: '修改组织失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '13': {// 功能编号
            app_biz_desp: '查找组织列表', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查找成功', desp: '查找成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查找组织列表失败', desp: '查找组织列表失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '14': {// 功能编号
            app_biz_desp: '查找组织类型', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '生成组织树成功', desp: '生成组织树成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '生成组织树失败', desp: '生成组织树失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '15': {// 功能编号
            app_biz_desp: '查找组织地区', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查找成功', desp: '查找成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查找组织地区失败', desp: '查找组织地区失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '16': {// 功能编号
            app_biz_desp: '生成组织树形目录', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查找成功', desp: '查找成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查找组织类型失败', desp: '查找组织类型失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '17': {// 功能编号
            app_biz_desp: '树形目录点击查询', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查询成功', desp: '查询节点及子节点信息成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查询失败', desp: '查询节点及子节点信息失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '17': {// 功能编号
            app_biz_desp: '查询父节点名称', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查询成功', desp: '查询父节点名称成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查询失败', desp: '查询父节点名称失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '18': {// 功能编号
            app_biz_desp: '查询当前用户角色', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查询成功', desp: '查询当前用户角色成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查询失败', desp: '查询当前用户角色失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        }
    }
};

module.exports = desp;
