// gmdf/lib/web/ecl/param.js
var desp = {
    app_id: 'gmdf-menu',// 模块编码
    app_author: 'yanghang',// 模块开发人员，gitlib账号名
    app_issueid:'', //开发任务ID ，如 #128
    app_biz_map: {// 功能列表
        '10': {// 功能编号
            app_biz_desp: '参数列表', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查询参数列表数据成功', desp: '查询参数列表数据成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查询参数列表数据异常', desp: '查询参数列表数据异常'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '11': {// 功能编号
            app_biz_desp: '保存参数', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '保存参数信息成功', desp: '保存参数信息成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '保存参数信息数据异常', desp: '保存参数信息数据异常'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '12': {// 功能编号
            app_biz_desp: '保存参数类别', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '保存参数类别信息成功', desp: '保存参数类别信息成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '保存参数类别信息数据异常', desp: '保存参数类别信息数据异常'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },

        '13': {// 功能编号
            app_biz_desp: '修改参数', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '修改参数信息成功', desp: '修改参数信息成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '必须选中一条数据', desp: 'id不能为空'},
                5101: {sdesp: '修改参数信息失败', desp: '修改参数信息失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '14': {// 功能编号
            app_biz_desp: '修改参数类别', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '修改参数类别信息成功', desp: '修改参数类别信息成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '必须选中一条数据', desp: 'id不能为空'},
                5101: {sdesp: '修改参数类别信息失败', desp: '修改参数类别信息失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '15': {// 功能编号
            app_biz_desp: '获取参数类别列表', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查找成功', desp: '查找成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查找参数类别列表失败', desp: '查找参数类别列表失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '16': {// 功能编号
            app_biz_desp: '获取系统列表', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查找成功', desp: '查找系统列表成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查找系统列表失败', desp: '查找系统列表失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '17': {// 功能编号
            app_biz_desp: '生成参数类别树形目录', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '生成菜单树成功', desp: '生成菜单树成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '生成菜单树失败', desp: '生成菜单树失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '18': {// 功能编号
            app_biz_desp: '点击参数类别改变数据表格内容', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查找成功', desp: '查询节点及子节点信息成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查询失败', desp: '查询节点及子节点信息失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '19': {// 功能编号
            app_biz_desp: '切换系统', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查找成功', desp: '查询选中系统菜单信息成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查询失败', desp: '查询选中系统菜单信息失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '20': {// 功能编号
            app_biz_desp: '获取类别名字', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '获取类别名字成功', desp: '获取类别名字成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '获取类别名字失败', desp: '获取类别名字失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
    }
};

module.exports = desp;
