// gmdf/lib/web/ecl/role.js
var desp = {
    app_id: 'gmdf-role',// 模块编码
    app_author: 'yanghang',// 模块开发人员，gitlib账号名
    app_issueid:'', //开发任务ID ，如 #128
    app_biz_map: {// 功能列表
        '10': {// 功能编号
            app_biz_desp: '角色列表', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查询角色列表数据成功', desp: '查询角色列表数据成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查询角色列表数据异常', desp: '查询角色列表数据异常'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '11': {// 功能编号
            app_biz_desp: '保存角色', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '保存角色信息成功', desp: '保存角色信息成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '保存角色信息异常', desp: '保存角色信息异常'},
                5101: {sdesp: '有未填入的选项', desp: '有未填入的选项'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '12': {// 功能编号
            app_biz_desp: '修改角色', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '修改角色信息成功', desp: '修改角色信息成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '必须选中一条数据', desp: 'id不能为空'},
                5101: {sdesp: '有未填入的选项', desp: '有未填入的选项'},
                5102: {sdesp: '修改角色信息失败', desp: '修改角色信息失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '13': {// 功能编号
            app_biz_desp: '获取系统列表', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查找成功', desp: '查找系统列表成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查找系统列表失败', desp: '查找系统列表失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '14': {// 功能编号
            app_biz_desp: '获取角色目录列表', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '生成角色树成功', desp: '生成角色树成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '生成角色树失败', desp: '生成角色树失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '15': {// 功能编号
            app_biz_desp: '获取拥有角色的菜单列表', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '生成角色树成功', desp: '生成带选中情况的角色树成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '生成角色树失败', desp: '生成带选中情况的角色树失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '16': {// 功能编号
            app_biz_desp: '保存权限到权限列表', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '菜单授权成功', desp: '菜单授权保存成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '菜单授权失败', desp: '菜单授权保存失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '17': {// 功能编号
            app_biz_desp: '保存权限到权限列表', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '系统切换成功', desp: '系统切换成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '系统切换失败', desp: '系统切换失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        }
    }
};

module.exports = desp;
