// gmdf/lib/web/ecl/user.js
var desp = {
    app_id: 'gmdf-user',// 模块编码
    app_author: 'yangjingya',// 模块开发人员，gitlib账号名
    app_issueid:'', //开发任务ID ，如 #128
    app_biz_map: {// 功能列表
        '10': {// 功能编号
            app_biz_desp: '用户列表', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查询用户列表数据成功', desp: '查询用户列表数据成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查询用户列表数据异常', desp: '查询用户列表数据异常'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '11': {// 功能编号
            app_biz_desp: '保存用户', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '保存用户成功', desp: '保存用户成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '新建用户失败', desp: '新建用户失败'},
                5101: {sdesp: '保存用户数据异常', desp: '保存用户数据异常'},
                5102: {sdesp: '保存角色数据异常', desp: '保存角色数据异常'},
                5103: {sdesp: '输入数据有空缺', desp: '输入数据有空缺'},
                5104: {sdesp: '所在组织机构有误', desp: '所在组织机构有误'},
                5105: {sdesp: '账户已存在', desp: '账户已被使用'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '12': {// 功能编号
            app_biz_desp: '修改用户', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '修改用户成功', desp: '修改用户成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '修改用户失败', desp: '修改用户失败'},
                5101: {sdesp: '必须选中一条数据', desp: 'id不能为空'},
                5102: {sdesp: '用户角色不能为空', desp: '用户角色不能为空'},
                5103: {sdesp: '修改用户数据失败', desp: '修改用户数据失败'},
                5104: {sdesp: '修改角色数据失败', desp: '修改角色数据失败'},
                5105: {sdesp: '所在组织机构有误', desp: '所在组织机构有误'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '13': {// 功能编号
            app_biz_desp: '查找系统列表', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查找成功', desp: '查找成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查找系统列表失败', desp: '查找系统列表失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '14': {// 功能编号
            app_biz_desp: '查找角色列表', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '查找成功', desp: '查找成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '查找用户列表失败', desp: '查找用户列表失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '15': {// 功能编号
            app_biz_desp: '重置密码', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid:'', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '重置密码成功,默认密码为：登录账号@cmcc', desp: '重置密码成功，默认密码为：登录账号@cmcc'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '必须选中一条数据', desp: 'id不能为空'},
                5101: {sdesp: '重置密码失败', desp: '重置密码失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '16': {// 功能编号
            app_biz_desp: '获取人员类别', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid: '', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '获取人员类别列表成功', desp: '获取人员类别列表成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '获取人员类别列表失败', desp: '获取人员类别列表失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '17': {// 功能编号
            app_biz_desp: '查询组织列表', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid: '', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '获取组织列表成功', desp: '获取组织列表成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '获取组织列表失败', desp: '获取组织列表失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '18': {// 功能编号
            app_biz_desp: '查询当前用户角色', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid: '', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '获取当前用户角色成功', desp: '获取当前用户角色成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '获取当前用户角色失败', desp: '获取当前用户角色失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '19': {// 功能编号
            app_biz_desp: '获取当前用户组织名称', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid: '', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '获取当前用户组织名称成功', desp: '获取当前用户组织名称成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '获取当前用户组织名称失败', desp: '获取当前用户组织名称失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        },
        '20': {// 功能编号
            app_biz_desp: '切换系统，获取当前系统用户人员信息', // 功能描述
            app_biz_modifier: '', // 功能修改人（非必填）
            app_issueid: '', //开发任务ID ，如 #128
            app_biz_code_map: {
                //-------------------调试和info代码---------------------------
                1101: {sdesp: '我的debug1', desp: '我的debug1'},
                //-------------------成功的代码---------------------------
                2000: {sdesp: '切换成功', desp: '获取当前系统用户人员信息成功'},
                //-------------------路由层代码51xx---------------------------
                5100: {sdesp: '切换失败', desp: '获取当前系统用户人员信息失败'},
                //-------------------业务层/JOB/其他服务代码52xx---------------------------
            }
        }
    }
};

module.exports = desp;
